### PR TITLE
DynamicLinking: Flesh out some more of the ABI details

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: DynamicLinking.md [hunk 1]
```diff
@@ -41,35 +41,66 @@ entire contents.
...
+
+(Note: the WebAssembly `start` function is not sufficient in all cases due to
+reentrancy issues, i.e., the `start` function is called as the module is being
+instantiated, before it returns its exports, so the outside cannot yet call into
+the module.)
+
+### Imports
+
+Functions are directly imported from the `env` module (e.g.
+`env.enternal_func `).  Data addresses and function addresses are imported as
```
Comment (kripken) [inline DynamicLinking.md]: extra space after `_func`

File: DynamicLinking.md [hunk 2]
```diff
@@ -41,35 +41,66 @@ entire contents.
...
+
+### Exports
+
+Functions are directly exported as WebAssembly function exports.  Exported
+addresses (i.e., exported memory locations or exported table locations) are
+exported as i32 WebAssembly globals.  However since exports are static, modules
+connect export the final relocated addresses (i.e. they cannot add
+`__memory_base` before exporting). Thus, the exported address is before
+relocation; the loader, which knows `__memory_base`, can then calculate the
+final relocated address.
```
Comment (kripken) [inline DynamicLinking.md]: Might be worth mentioning that exports are callable both from JS and from wasm, so they must be legalized? Or maybe something to keep thinking about later.

File: DynamicLinking.md [hunk 3]
```diff
@@ -41,35 +41,66 @@ entire contents.
...
-   before relocation; the loader, which knows `__memory_base`, can then
-   calculate the proper relocated and final address of those addresses.
+   outside. If it does so, it should import `env.__table_base`, an `i32`
+   WebAssembly global, in which the loader will provide the start of the region
+   in the table which has been reserved for this module, as described earlier.
+
+### Relocations
+
+WebAssembly dynamic libraies do not require relocations in the code section.
+This allows for streaming complication and better code sharing, and reduces the
```
Comment (pepyakin) [inline DynamicLinking.md]: typo: complication -> compilation

File: DynamicLinking.md [hunk 4]
```diff
@@ -41,35 +41,66 @@ entire contents.
...
+
+### Imports
+
+Functions are directly imported from the `env` module (e.g.
+`env.enternal_func `).  Data addresses and function addresses are imported as
+functions that return the address.  This is because the final addresse of given
+symbol might not be known until all modules are initialized.  These functions
+are named with the `g$` prefix. For example `env.g$goo` can be imported and
+used to access the address of `foo`.  In the futre this scheme could be
+replaced with importing of mutable globals.
```
Comment (pepyakin) [inline DynamicLinking.md]: Why mutable globals can not be used now?

Comment (sbc100) [reply DynamicLinking.md]: importing/exporting of mutable globals was added more recently and IIUC is not yet available in all 4 major browsers.

Also, I'm kind of documenting how things work today, and I will be added another section which is how they might work in the future.

File: DynamicLinking.md (hunk 2 above)
Comment (sbc100) [reply DynamicLinking.md]: I'm not sure how much JS-specific stuff we want in this doc.  Perhaps a separate section at the end?

**Pull request merged**